### DISCUSSION TITLE
Why is `typing.Final` not usable if re-exported by some other modules?

### DISCUSSION THREAD

**Discussion by Azureblade3808** [microsoft/pyright#4845] [Q&A] [answered]
I was trying to move the most commonly used types into one file, in order to reduce unnecessary runtime imports. Finally I came up with something like --
```python
# _t.pyi

from typing import (
    Any as Any,
    Final as Final,
    Iterable as Iterable,
    Mapping as Mapping,
    NoReturn as NoReturn,
    Sequence as Sequence,
)
from typing_extensions import (
    Never as Never,
    ParamSpec,
    TypeAlias as TypeAlias,
)

from pandas import (
    DataFrame as DataFrame,
    Series as Series,
)

P = ParamSpec("P")
```

For usage, it will be like --
```python
# some.py

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from . import _t  # pyright: ignore[reportMissingModuleSource]

def get_df() -> _t.DataFrame: ...
```

The `_t` module doesn't exist in runtime. By using it with `TYPE_CHECKING`, I can avoid some heavy imports (e.g. matplotlib) before they are truly needed. And a `_t.` prefix in the types explicitly indicates that the types cannot be used in runtime calls, so that coding errors like --
```python
if isinstance(x, _t.DataFrame): ...
```
can be easily picked.

I've placed `Final` here because it's only ever used in type checking, but found that code below doesn't work --
```python
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from . import _t  # pyright: ignore[reportMissingModuleSource]

final_value: _t.Final = 0
```
I think this is because `Final` is more like a reserved keyword and needs to be directly imported from `typing`/`typing_extensions` to be recognized. But by contrast, I've found that `TypeVar`, `Generic` and `Protocol` are all usable when re-exported.

For what I've checked:
Usable (re-exported) | pyright | mypy
-|-|-
Any | Yes | Yes
Final | No | No
Generic | Yes | No
Never | Yes | Yes
NoReturn | Yes | Yes
Protocol | Yes | No
TypeVar | Yes | No
final | Yes | No
* By "usable", it means that the re-exported version works the same with the original version.

Here are my questions:
1) Is there a reason behind these behaviors (which I think are inconsistent)? Or should I think them as UBs and just avoid using any  re-exported versions of these keyword-likes?
1) Is it safe to use re-exported versions of protocols such as `Iterable`, `Sequence`?

Edit
----
`final` (lowercase) is usable when re-exported with pyright.

**Comment by erictraut** [comment, marked as answer]
Yes, in any other language `Final` would be a reserved keyword and part of the language itself. Python has an aversion to adding typing-related keywords, so we're forced to use these hacky mechanisms to add functionality like `Final`. These need to be understood early in the static analysis process — long before types can be evaluated, so aliasing them across files doesn't work. 

The following special forms cannot be re-exported: `Final`, `ClassVar`, and `InitVar` in pyright.

You indicated that `final` (lowercase) does not work when re-exported, but it should work fine in pyright.

**Reply by Azureblade3808** [reply]
Oh my mistake. I confirmed that `final` does work when re-exported with pyright.
